### PR TITLE
Add Replace

### PR DESCRIPTION
**Pull request by eliascarv**
This PR implements Replace tranformation. The `Replace(oldᵢ => newᵢ)` transformations replaces all occurrences of `oldᵢ` with `newᵢ` in the table.

Note: If the type of `old` is different from the type of `new`, the column type is changed to a type common to the new column elements, without promotion.
Example:
```julia-repl
julia> using TableTransforms, DataFrames

julia> df = DataFrame(a = [2, 1, 2], b = [1, 1, 1])
3×2 DataFrame       
 Row │ a      b     
     │ Int64  Int64 
─────┼──────────────
   1 │     2      1 
   2 │     1      1 
   3 │     2      1 

julia> T = Replace(1 => 1.5)
Replace{Int64, Float64}(IdDict(1 => 1.5))

julia> T(df)
3×2 DataFrame
 Row │ a     b       
     │ Real  Float64 
─────┼───────────────
   1 │  2        1.5
   2 │  1.5      1.5
   3 │  2        1.5
```

### CODE REVIEW

File: src/TableTransforms.jl [hunk 1]
```diff
@@ -39,6 +39,7 @@ export
   Coalesce,
   Identity,
   Center,
+  Replace,
```
Comment (juliohm) [inline src/TableTransforms.jl:42]: Move it to before Identity. Everything below Identity is more "numerical".

File: src/transforms/replace.jl [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+"""
+    Replace(old₁ => new₁, old₂ => new₂, ..., oldₙ => newₙ)
+
+Replaces all occurrences of `oldᵢ` with `newᵢ` in the table.
+"""
+struct Replace{K,V} <: Colwise
+  oldnew::IdDict{K,V}
```
Comment (juliohm) [inline src/transforms/replace.jl:10]: Should we call this pairs for consistency with other transforms? Also, first time I see IdDict, interesting finding.

**Review by juliohm — CHANGES_REQUESTED** [review]

File: src/transforms/replace.jl (hunk 2 above)
Comment (eliascarv) [reply src/transforms/replace.jl:10]: OK.
IdDict is used for this reason:
```julia-repl
julia> Dict((1 => 3, true => false))
Dict{Integer, Integer} with 1 entry:
  true => false

julia> IdDict((1 => 3, true => false))
IdDict{Integer, Integer} with 2 entries:
  1    => 3
  true => false
```

File: src/TableTransforms.jl (hunk 1 above)
Comment (eliascarv) [reply src/TableTransforms.jl:42]: OK

Comment (eliascarv) [reply src/TableTransforms.jl]: Made in [0159ff2](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/0159ff21165a22578f40d5d28e94e37a44c31b66)

**Review by juliohm — COMMENTED** [review]
Need to resolve newly added conflicts with master.

File: src/TableTransforms.jl [hunk 3]
```diff
@@ -37,6 +37,7 @@ export
   DropMissing,
   Rename,
   Coalesce,
+  Replace,
```
Comment (juliohm) [inline src/TableTransforms.jl:40]: Move to before Coalesce

File: src/transforms.jl [hunk 4]
```diff
@@ -215,6 +215,7 @@ include("transforms/select.jl")
 include("transforms/filter.jl")
 include("transforms/rename.jl")
 include("transforms/coalesce.jl")
+include("transforms/replace.jl")
```
Comment (juliohm) [inline src/transforms.jl:218]: Same here.

File: src/transforms/replace.jl [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    Replace(old₁ => new₁, old₂ => new₂, ..., oldₙ => newₙ)
+
+Replaces all occurrences of `oldᵢ` with `newᵢ` in the table.
```
Comment (juliohm) [inline src/transforms/replace.jl:8]:
```suggestion
Replaces `oldᵢ` value with `newᵢ` value in the table.
```

File: test/transforms.jl [hunk 6]
```diff
@@ -673,6 +673,89 @@
...
+    tₒ = revert(T, n, c)
+    @test t == tₒ
+
+    # table schema after apply and revert
+    T = Replace(1 => 1.5, 5 => 5.5, 4 => true)
+    n, c = apply(T, t)
+    tₒ = revert(T, n, c)
+    ttypes = Tables.schema(t).types
+    ntypes = Tables.schema(n).types
+    @test ntypes[1] == typejoin(typeof.(n.a)...)
```
Comment (juliohm) [inline test/transforms.jl:724]: Isn't this the eltype of n.a? Can we just use eltype?

**Review by juliohm — CHANGES_REQUESTED** [review]

File: test/transforms.jl (hunk 6 above)
Comment (eliascarv) [reply test/transforms.jl:724]: `n.a` has values with different types.
In this case:
```julia
a = Real[3, 2, 1.5, true, 5.5, 3]
typejoin(typeof.(a)...)
typejoin(Int, Int, Float64, Bool, Float64, Int)
Real
```
It's just to make sure the common type between column values ​​is correct.

File: src/TableTransforms.jl (hunk 3 above)
Comment (eliascarv) [reply src/TableTransforms.jl:40]: OK

File: src/transforms.jl (hunk 4 above)
Comment (eliascarv) [reply src/transforms.jl:218]: OK

File: src/transforms/replace.jl (hunk 5 above)
Comment (eliascarv) [reply src/transforms/replace.jl:8]: OK

Comment (eliascarv) [reply src/transforms/replace.jl]: Made in [6e3d31e](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/6e3d31efa6523dd7b8c4fb85c21d07ac9cc4e738)

File: src/transforms.jl (hunk 4 above)
Comment (eliascarv) [reply src/transforms.jl:217]: Made in [9c5c293](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/9c5c293c10346a2cb1eddf7f2bf11923779906de)

File: src/TableTransforms.jl (hunk 3 above)
Comment (eliascarv) [reply src/TableTransforms.jl:39]: Made in [9c5c293](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/9c5c293c10346a2cb1eddf7f2bf11923779906de)

File: test/transforms.jl (hunk 6 above)
Comment (juliohm) [reply test/transforms.jl:679]: Yes, but eltype(n.a) returns Real as well. What am I missing?

Comment (eliascarv) [reply test/transforms.jl:679]: `eltype(n.a)` could be `Number` or `Any`:
```julia
a = Any[3, 2, 1.5, true, 5.5, 3]
typejoin(typeof.(a)...)
typejoin(Int, Int, Float64, Bool, Float64, Int)
Real
```
This check is just to make sure the algorithm is working correctly.

File: src/transforms/replace.jl [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+isrevertible(::Type{<:Replace}) = true
+
+function colcache(transform::Replace, x)
+  olds = keys(transform.pairs)
+  inds = [findall(v -> v === old, x) .=> old for old in olds]
+  Dict(reduce(vcat, inds))
+end
+
+colapply(transform::Replace, x, c) =
+  map(v -> get(transform.pairs, v, v), x)
```
Comment (juliohm) [inline src/transforms/replace.jl:27]: Can we use a simple list comprehension here?

File: src/transforms/replace.jl [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+colrevert(::Replace, x, c) =
+  map(i -> get(c, i, x[i]), 1:length(x))
```
Comment (juliohm) [inline src/transforms/replace.jl:30]: Same here, a list comprehension can improve the readability of this line.

File: test/transforms.jl (hunk 6 above)
Comment (juliohm) [reply test/transforms.jl:679]: I don't understand, why we cannot use eltype in this test? What do we gain with typejoin(typeof.(...))?

**Review by juliohm — CHANGES_REQUESTED** [review]

File: test/transforms.jl (hunk 6 above)
Comment (eliascarv) [reply test/transforms.jl]: OK, i will change the tests.

Comment (eliascarv) [reply test/transforms.jl]: Made in [2f87e1d](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/2f87e1d78626a6982403f3de4ffd39eb107f4ab8)

File: src/transforms/replace.jl (hunk 7 above)
Comment (eliascarv) [reply src/transforms/replace.jl]: Made in [2f87e1d](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/2f87e1d78626a6982403f3de4ffd39eb107f4ab8)

File: src/transforms/replace.jl (hunk 8 above)
Comment (eliascarv) [reply src/transforms/replace.jl]: Made in [2f87e1d](https://github.com/JuliaML/TableTransforms.jl/pull/42/commits/2f87e1d78626a6982403f3de4ffd39eb107f4ab8)

File: src/transforms/replace.jl [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  olds = keys(transform.pairs)
+  inds = [findall(v -> v === old, x) .=> old for old in olds]
+  Dict(reduce(vcat, inds))
+end
+
+colapply(transform::Replace, x, c) =
+  [get(transform.pairs, xᵢ, xᵢ) for xᵢ in x]
+
+colrevert(::Replace, x, c) =
+  [get(c, i, x[i]) for i in 1:length(x)]
```
Comment (juliohm) [inline src/transforms/replace.jl:30]:
```suggestion
colapply(transform::Replace, x, c) = [get(transform.pairs, xᵢ, xᵢ) for xᵢ in x]

colrevert(transform::Replace, x, c) = [get(c, i, x[i]) for i in 1:length(x)]
```

File: src/transforms/replace.jl (hunk 7 above)
Comment (eliascarv) [reply src/transforms/replace.jl]: Which of these options do you prefer:
One line:
```julia
colapply(transform::Replace, x, c) = [get(transform.pairs, xᵢ, xᵢ) for xᵢ in x]
```
Two lines:
```julia
colapply(transform::Replace, x, c) = 
  [get(transform.pairs, xᵢ, xᵢ) for xᵢ in x]
```

**Review by juliohm — APPROVED** [review]
The IdDict trick was clever 💯

File: src/transforms/replace.jl (hunk 8 above)
Comment (eliascarv) [reply src/transforms/replace.jl]: Which of these options do you prefer:
One line:
```julia
colrevert(::Replace, x, c) = [get(c, i, x[i]) for i in 1:length(x)]
```
Two lines:
```julia
colrevert(::Replace, x, c) = 
  [get(c, i, x[i]) for i in 1:length(x)]
```

File: test/transforms.jl [hunk 10]
```diff
@@ -672,7 +755,7 @@
     @test ntypes[6] == Int
     @test ttypes == Tables.schema(tₒ).types    
   end
-
+
```
Comment (juliohm) [inline test/transforms.jl:758]:
```suggestion

```

**Comment by eliascarv** [comment]
Thanks @juliohm!

**Comment by eliascarv** [comment]
Can I merge when tests pass @juliohm?

**Comment by juliohm** [comment]
Yes, feel free to merge when all green tickets show up. 👍🏽

**Pull request merged**